File: output/python/tests/14_system_return_default.py
from typing import Any, Optional, List, Dict, Callable

class SystemReturnDefaultTestFrameEvent:
    def __init__(self, message: str, parameters):
        self._message = message
        self._parameters = parameters
        self._return = None


class SystemReturnDefaultTestCompartment:
    def __init__(self, state: str, parent_compartment = None):
        self.state = state
        self.state_args = {}
        self.state_vars = {}
        self.enter_args = {}
        self.exit_args = {}
        self.forward_event = None
        self.parent_compartment = parent_compartment

    def copy(self) -> 'SystemReturnDefaultTestCompartment':
        c = SystemReturnDefaultTestCompartment(self.state, self.parent_compartment)
        c.state_args = self.state_args.copy()
        c.state_vars = self.state_vars.copy()
        c.enter_args = self.enter_args.copy()
        c.exit_args = self.exit_args.copy()
        c.forward_event = self.forward_event
        return c


class SystemReturnDefaultTest:
    def __init__(self):
        self._state_stack = []
        self._return_value = None
        self.__compartment = SystemReturnDefaultTestCompartment("Start")
        self.__next_compartment = None
        __frame_event = SystemReturnDefaultTestFrameEvent("$>", None)
        self.__kernel(__frame_event)

    def __kernel(self, __e):
        # Route event to current state
        self.__router(__e)
        # Process any pending transition
        while self.__next_compartment is not None:
            next_compartment = self.__next_compartment
            self.__next_compartment = None
            # Exit current state
            exit_event = SystemReturnDefaultTestFrameEvent("<$", self.__compartment.exit_args)
            self.__router(exit_event)
            # Switch to new compartment
            self.__compartment = next_compartment
            # Enter new state (or forward event)
            if next_compartment.forward_event is None:
                enter_event = SystemReturnDefaultTestFrameEvent("$>", self.__compartment.enter_args)
                self.__router(enter_event)
            else:
                # Forward event to new state
                forward_event = next_compartment.forward_event
                next_compartment.forward_event = None
                if forward_event._message == "$>":
                    # Forwarding enter event - just send it
                    self.__router(forward_event)
                else:
                    # Forwarding other event - send $> first, then forward
                    enter_event = SystemReturnDefaultTestFrameEvent("$>", self.__compartment.enter_args)
                    self.__router(enter_event)
                    self.__router(forward_event)

    def __router(self, __e):
        state_name = self.__compartment.state
        handler_name = f"_state_{state_name}"
        handler = getattr(self, handler_name, None)
        if handler:
            handler(__e)

    def __transition(self, next_compartment):
        self.__next_compartment = next_compartment

    def handler_sets_value(self) -> str:
        self._return_value = None
        __e = SystemReturnDefaultTestFrameEvent("handler_sets_value", None)
        self.__kernel(__e)
        return self._return_value

    def handler_no_return(self) -> str:
        self._return_value = None
        __e = SystemReturnDefaultTestFrameEvent("handler_no_return", None)
        self.__kernel(__e)
        return self._return_value

    def get_count(self) -> int:
        self._return_value = None
        __e = SystemReturnDefaultTestFrameEvent("get_count", None)
        self.__kernel(__e)
        return self._return_value

    def _state_Start(self, __e):
        if __e._message == "$>":
            self.__compartment.state_vars["count"] = 0
        elif __e._message == "get_count":
            self._return_value = self.__compartment.state_vars["count"]
            __e._return = self._return_value
            return
        elif __e._message == "handler_no_return":
            # Does not set return - should return None
            self.__compartment.state_vars["count"] = self.__compartment.state_vars["count"] + 1
        elif __e._message == "handler_sets_value":
            self._return_value = "set_by_handler"
            __e._return = self._return_value
            return


def main():
    print("=== Test 14: System Return Default Behavior ===")
    s = SystemReturnDefaultTest()

    # Test 1: Handler explicitly sets return value
    result = s.handler_sets_value()
    assert result == "set_by_handler", f"Expected 'set_by_handler', got '{result}'"
    print(f"1. handler_sets_value() = '{result}'")

    # Test 2: Handler does NOT set return - should return None
    result = s.handler_no_return()
    assert result is None, f"Expected None, got '{result}'"
    print(f"2. handler_no_return() = {result}")

    # Test 3: Verify handler was called (side effect check)
    count = s.get_count()
    assert count == 1, f"Expected count=1, got {count}"
    print(f"3. Handler was called, count = {count}")

    # Test 4: Call again to verify idempotence
    result = s.handler_no_return()
    assert result is None, f"Expected None again, got '{result}'"
    count = s.get_count()
    assert count == 2, f"Expected count=2, got {count}"
    print(f"4. Second call: result={result}, count={count}")

    print("PASS: System return default behavior works correctly")

if __name__ == '__main__':
    main()

File: output/javascript/tests/14_system_return_default.js

// NOTE: Default return value syntax (method(): type = default) not yet implemented.
// This test validates behavior when handler does not set @@:return.

export class SystemReturnDefaultTestFrameEvent {
    _message;
    _parameters;

    constructor(message, parameters) {
        this._message = message;
        this._parameters = parameters;
    }
}


export class SystemReturnDefaultTestFrameContext {
    event;
    _return;
    _data;

    constructor(event, default_return) {
        this.event = event;
        this._return = default_return;
        this._data = {  };
    }
}


export class SystemReturnDefaultTestCompartment {
    state;
    state_args;
    state_vars;
    enter_args;
    exit_args;
    forward_event;
    parent_compartment;

    constructor(state, parent_compartment = null) {
        this.state = state;
        this.state_args = {  };
        this.state_vars = {  };
        this.enter_args = {  };
        this.exit_args = {  };
        this.forward_event = null;
        this.parent_compartment = parent_compartment;
    }

    copy() {
        const c = new SystemReturnDefaultTestCompartment(this.state, this.parent_compartment);
        c.state_args = {...this.state_args};
        c.state_vars = {...this.state_vars};
        c.enter_args = {...this.enter_args};
        c.exit_args = {...this.exit_args};
        c.forward_event = this.forward_event;
        return c;
    }
}


export class SystemReturnDefaultTest {
    _state_stack;
    __compartment;
    __next_compartment;
    _context_stack;

    constructor() {
        this._state_stack = [];
        this._context_stack = [];
        this.__compartment = new SystemReturnDefaultTestCompartment("Start");
        this.__next_compartment = null;
        const __frame_event = new SystemReturnDefaultTestFrameEvent("$>", null);
        this.__kernel(__frame_event);
    }

    __kernel(__e) {
        // Route event to current state
        this.__router(__e);
        // Process any pending transition
        while (this.__next_compartment !== null) {
            const next_compartment = this.__next_compartment;
            this.__next_compartment = null;
            // Exit current state
            const exit_event = new SystemReturnDefaultTestFrameEvent("<$", this.__compartment.exit_args);
            this.__router(exit_event);
            // Switch to new compartment
            this.__compartment = next_compartment;
            // Enter new state (or forward event)
            if (next_compartment.forward_event === null) {
                const enter_event = new SystemReturnDefaultTestFrameEvent("$>", this.__compartment.enter_args);
                this.__router(enter_event);
            } else {
                // Forward event to new state
                const forward_event = next_compartment.forward_event;
                next_compartment.forward_event = null;
                if (forward_event._message === "$>") {
                    // Forwarding enter event - just send it
                    this.__router(forward_event);
                } else {
                    // Forwarding other event - send $> first, then forward
                    const enter_event = new SystemReturnDefaultTestFrameEvent("$>", this.__compartment.enter_args);
                    this.__router(enter_event);
                    this.__router(forward_event);
                }
            }
        }
    }

    __router(__e) {
        const state_name = this.__compartment.state;
        const handler_name = `_state_${state_name}`;
        const handler = this[handler_name];
        if (handler) {
            handler.call(this, __e);
        }
    }

    __transition(next_compartment) {
        this.__next_compartment = next_compartment;
    }

    handler_sets_value() {
        const __e = new SystemReturnDefaultTestFrameEvent("handler_sets_value", null);
        const __ctx = new SystemReturnDefaultTestFrameContext(__e, null);
        this._context_stack.push(__ctx);
        this.__kernel(__e);
        return this._context_stack.pop()._return;
    }

    handler_no_return() {
        const __e = new SystemReturnDefaultTestFrameEvent("handler_no_return", null);
        const __ctx = new SystemReturnDefaultTestFrameContext(__e, null);
        this._context_stack.push(__ctx);
        this.__kernel(__e);
        return this._context_stack.pop()._return;
    }

    get_count() {
        const __e = new SystemReturnDefaultTestFrameEvent("get_count", null);
        const __ctx = new SystemReturnDefaultTestFrameContext(__e, null);
        this._context_stack.push(__ctx);
        this.__kernel(__e);
        return this._context_stack.pop()._return;
    }

    _state_Start(__e) {
        // HSM: Navigate to this state's compartment for state var access
        let __sv_comp = this.__compartment;
        while (__sv_comp !== null && __sv_comp.state !== "Start") {
            __sv_comp = __sv_comp.parent_compartment;
        }
        if (__e._message === "$>") {
            if (!("count" in __sv_comp.state_vars)) {
                __sv_comp.state_vars["count"] = 0;
            }
        } else if (__e._message === "get_count") {
            this._context_stack[this._context_stack.length - 1]._return = __sv_comp.state_vars["count"];
            return;;
        } else if (__e._message === "handler_no_return") {
            // Does not set return - should return null/undefined
            __sv_comp.state_vars["count"] = __sv_comp.state_vars["count"] + 1;
        } else if (__e._message === "handler_sets_value") {
            this._context_stack[this._context_stack.length - 1]._return = "set_by_handler";
            return;;
        }
    }
}

function main() {
    console.log("=== Test 14: System Return Default Behavior ===");
    const s = new SystemReturnDefaultTest();

    // Test 1: Handler explicitly sets return value
    const result1 = s.handler_sets_value();
    if (result1 !== "set_by_handler") {
        throw new Error(`Expected 'set_by_handler', got '${result1}'`);
    }
    console.log(`1. handler_sets_value() = '${result1}'`);

    // Test 2: Handler does NOT set return - should return null/undefined
    const result2 = s.handler_no_return();
    if (result2 !== null && result2 !== undefined) {
        throw new Error(`Expected null/undefined, got '${result2}'`);
    }
    console.log(`2. handler_no_return() = ${result2}`);

    // Test 3: Verify handler was called (side effect check)
    let count = s.get_count();
    if (count !== 1) {
        throw new Error(`Expected count=1, got ${count}`);
    }
    console.log(`3. Handler was called, count = ${count}`);

    // Test 4: Call again to verify idempotence
    const result4 = s.handler_no_return();
    if (result4 !== null && result4 !== undefined) {
        throw new Error(`Expected null/undefined again, got '${result4}'`);
    }
    count = s.get_count();
    if (count !== 2) {
        throw new Error(`Expected count=2, got ${count}`);
    }
    console.log(`4. Second call: result=${result4}, count=${count}`);

    console.log("PASS: System return default behavior works correctly");
}

main();
